File: dp/c/pimpl.c
// Pimpl (Opaque Pointer) Pattern in C
// Hide implementation behind a forward-declared struct

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// --- "header" portion: user sees only opaque pointer ---
typedef struct Engine Engine;

Engine* engine_create(int cylinders, int hp);
void engine_start(const Engine* e);
void engine_stop(const Engine* e);
const char* engine_description(const Engine* e);
void engine_destroy(Engine* e);

// --- "implementation" portion ---
struct Engine {
    int cylinders;
    int horsepower;
    char desc[64];
};

Engine* engine_create(int cylinders, int hp) {
    Engine* e = malloc(sizeof(Engine));
    e->cylinders = cylinders;
    e->horsepower = hp;
    snprintf(e->desc, 64, "%dcyl %dhp engine", cylinders, hp);
    return e;
}

void engine_start(const Engine* e) {
    printf("[Engine] Starting %dcyl (%dhp)\n", e->cylinders, e->horsepower);
}

void engine_stop(const Engine* e) {
    printf("[Engine] Stopping\n");
}

const char* engine_description(const Engine* e) { return e->desc; }

void engine_destroy(Engine* e) { free(e); }

int main(void) {
    Engine* e = engine_create(4, 150);
    printf("%s\n", engine_description(e));
    engine_start(e);
    engine_stop(e);
    engine_destroy(e);
    return 0;
}

File: dp/rust/src/pimpl.rs
// Pimpl (Pointer to Implementation) Pattern in Rust
// Hide implementation details behind a Box<dyn Trait> or opaque struct

// In Rust, this is idiomatically achieved with trait objects or modules

mod engine {
    // Private implementation — not exposed to users
    struct EngineImpl {
        cylinders: u32,
        horsepower: u32,
    }

    // Public interface
    pub struct Engine {
        inner: Box<EngineImpl>,
    }

    impl Engine {
        pub fn new(cylinders: u32, horsepower: u32) -> Self {
            Engine {
                inner: Box::new(EngineImpl { cylinders, horsepower }),
            }
        }

        pub fn description(&self) -> String {
            format!("{}cyl {}hp engine", self.inner.cylinders, self.inner.horsepower)
        }

        pub fn start(&self) {
            println!("[Engine] Starting {} ({}hp)", self.inner.cylinders, self.inner.horsepower);
        }

        pub fn stop(&self) {
            println!("[Engine] Stopping");
        }
    }
}

// Trait-based pimpl (more idiomatic Rust)
trait Renderer {
    fn render(&self, content: &str);
    fn clear(&self);
}

struct OpenGLRenderer;
impl Renderer for OpenGLRenderer {
    fn render(&self, content: &str) {
        println!("[OpenGL] Rendering: {}", content);
    }
    fn clear(&self) {
        println!("[OpenGL] Clear buffer");
    }
}

struct Widget {
    renderer: Box<dyn Renderer>,  // pimpl via trait object
    name: String,
}

impl Widget {
    fn new(name: &str, renderer: Box<dyn Renderer>) -> Self {
        Widget { renderer, name: name.to_string() }
    }

    fn draw(&self) {
        self.renderer.render(&self.name);
    }
}

fn main() {
    // Opaque struct pimpl
    println!("=== Opaque Struct ===");
    let engine = engine::Engine::new(4, 150);
    println!("{}", engine.description());
    engine.start();
    engine.stop();

    // Trait object pimpl
    println!("\n=== Trait Object ===");
    let widget = Widget::new("Button", Box::new(OpenGLRenderer));
    widget.draw();
}
